### PR TITLE
Implements Jupyter Project

### PR DESCRIPTION
**Pull request by AlexCatarino**
With this new project, we are able to use Jupyter with QuantConnect data for quantitative analysis

### CODE REVIEW

File: DockerfileJupyter [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#	LEAN Jupyter Docker Container 20170802
+#	
+
+# Use base system for cleaning up wayward processes
+FROM quantconnect/lean:foundation
+
+MAINTAINER QuantConnect <contact@quantconnect.com>
+
+# Install Tini
+RUN wget --quiet https://github.com/krallin/tini/releases/download/v0.10.0/tini && \
```
Comment (andrewhart098) [inline DockerfileJupyter]: Possibly we don't need tini? From the tini project page 

> NOTE:  If you are using Docker 1.13 or greater, Tini is included in Docker itself. This includes all versions of Docker CE. To enable Tini, just pass the --init flag to docker run.

File: DockerfileJupyter [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    echo "1361527f39190a7338a0b434bd8c88ff7233ce7b9a4876f3315c22fce7eca1b0 *tini" | sha256sum -c - && \
+    mv tini /usr/local/bin/tini && \
+    chmod +x /usr/local/bin/tini
+
+#Install Jupyter
+RUN pip install jupyter
+	
+# Install Python.NET
+RUN apt-get update && apt-get install -y libglib2.0-dev \
```
Comment (andrewhart098) [inline DockerfileJupyter]: Isn't python.Net included in the quantconnect/lean:foundation docker file?

File: DockerfileJupyter [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+RUN rm -irf ${LEAN}/*Python*
+
+WORKDIR $WORK
+
+ENTRYPOINT [ "/usr/local/bin/tini", "--" ]
+CMD jupyter notebook --ip='*' --port=8888 --no-browser --allow-root
+
+# Usage: 
+# docker build -t quantconnect/lean:foundation -f DockerfileLeanFoundation .
+# docker build -t quantconnect/lean:jupyter -f DockerfileJupyter .
```
Comment (andrewhart098) [inline DockerfileJupyter]: Why not quantconnect/jupyter:latest?

File: DockerfileJupyter [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+# docker run -v "(absolute to your data folder):/root/Lean/Data" quantconnect/lean:jupyter
```
Comment (andrewhart098) [inline DockerfileJupyter]: May want to recommend that docker mounts the data folder in readonly mode so,
`docker run -v (absolute to your data folder):/root/Lean/Data:ro quantconnect/lean:jupyter `

File: DockerfileJupyter (hunk 2 above)
Comment (AlexCatarino) [inline DockerfileJupyter]: In the foundation, we have a pythonnet installation that only works in Lean. For instance, if you open python and type "import clr", it says the package is not installed.
We could change the foundation to include this installation method that is more general.

File: DockerfileJupyter (hunk 4 above)
Comment (andrewhart098) [inline DockerfileJupyter]: The projects in Examples won't work with this run command.  The problem is that the example projects live in the path `~/Lean/Jupyter/bin/Release/Examples` and the data folder is specified in the config as `../../../Data`. Therefore, when in one of the Example projects, the absolute path where Lean is looking for data is `~/Lean/Jupyter/Data` when this run command mounts it at `~/Lean/Data`. The data path should probably be configured to an absolute path in Lean

File: DockerfileJupyter (hunk 3 above)
Comment (AlexCatarino) [inline DockerfileJupyter]: I thought that since it is part of Lean, it made sense.

File: DockerfileJupyter (hunk 4 above)
Comment (AlexCatarino) [inline DockerfileJupyter]: Maybe we just move them to the parent directory.

File: Jupyter/QuantBook.cs [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+        }
+
+        /// <summary>
+        /// Get the Lean <see cref = "Symbol"/> object given its ticker, security type and market.
+        /// </summary>
+        /// <param name="ticker">The asset symbol name</param>
+        /// <param name="type">The asset security type</param>
+        /// <param name="market">The asset market</param>
+        /// <returns><see cref = "Symbol"/> object wrapped in a <see cref = "PyObject"/></returns>
+        public PyObject get_symbol(string ticker, string type = "Equity", string market = null)
```
Comment (StefanoRaggi) [inline Jupyter/QuantBook.cs]: This method should be renamed to `GetSymbol`

File: Jupyter/QuantBook.cs [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+
+            return SymbolCache.GetSymbol(ticker).ToPython();
+        }
+
+        /// <summary>
+        /// Gets symbol information from a list of tickers
+        /// </summary>
+        /// <param name="pyObject">The tickers to retrieve information for</param>
+        /// <returns>A pandas.DataFrame containing the symbol information</returns>
+        public PyObject print_symbols(PyObject pyObject)
```
Comment (StefanoRaggi) [inline Jupyter/QuantBook.cs]: This method should be renamed to `PrintSymbols`

File: Jupyter/QuantBook.cs [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+                    DataNormalizationMode = dataNormalizationMode,
+                    IncludeExtendedMarketHours = extendedMarket,
+                    DataType = symbol.ID.SecurityType == SecurityType.Equity ? typeof(TradeBar) : typeof(QuoteBar)
+                };
+            });
+
+            return CreateDataFrame(requests, _historyProvider.GetHistory(requests, TimeZones.NewYork), selector);
+        }
+
+        public PyObject GetFundamental(PyObject pyObject, string selector, DateTime? start = null, DateTime? end = null)
```
Comment (StefanoRaggi) [inline Jupyter/QuantBook.cs]: XML comment is missing for this public method

File: Jupyter/packages.config [hunk 8]
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="NodaTime" version="1.3.1" targetFramework="net452" />
```
Comment (StefanoRaggi) [inline Jupyter/packages.config]: This reference should be updated to version `1.3.4`

**Comment by idcrisis** [comment]
Thank you.

Wonder if exchange data integrity can be signed. 

Also multiplicative UDP broadcast of quotes free. Hourly especially.

File: Jupyter/AlgorithmAnalysis.cs [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        {
+            try
+            {
+                using (Py.GIL())
+                {
+                    _pandas = Py.Import("pandas");
+                }
+
+                // Set the data-folder
+                Config.Set("data-folder", "/root/Lean/Data");
```
Comment (jaredbroad) [inline Jupyter/AlgorithmAnalysis.cs]: Don't hard code data folder, should be configured.

**Review by 2aspace — COMMENTED** [review]
Je ill

**Pull request merged**